ingestion/src/metadata/ingestion/source/database/datalake.py

```diff
@@ -0,0 +1,325 @@
+    OpenMetadataConnection,
+)
+from metadata.generated.schema.entity.services.databaseService import DatabaseService
+from metadata.generated.schema.metadataIngestion.databaseServiceMetadataPipeline import (
+    DatabaseServiceMetadataPipeline,
+)
+from metadata.generated.schema.metadataIngestion.workflow import (
+    Source as WorkflowSource,
+)
+from metadata.generated.schema.type.entityReference import EntityReference
+from metadata.ingestion.api.common import Entity
+from metadata.ingestion.api.source import InvalidSourceException, Source, SourceStatus
+from metadata.ingestion.models.ometa_table_db import OMetaDatabaseAndTable
+from metadata.ingestion.ometa.ometa_api import OpenMetadata
+from metadata.ingestion.source.database.common_db_source import SQLSourceStatus
+from metadata.utils.column_type_parser import ColumnTypeParser
+from metadata.utils.connections import get_connection
+from metadata.utils.filters import filter_by_table
+from metadata.utils.logger import ingestion_logger
+
+logger = ingestion_logger()
+
+
+class DatalakeSource(Source[Entity]):
+    def __init__(self, config: WorkflowSource, metadata_config: OpenMetadataConnection):
+        super().__init__()
+        self.status = SQLSourceStatus()
+
+        self.config = config
+        self.source_config: DatabaseServiceMetadataPipeline = (
+            self.config.sourceConfig.config
+        )
+        self.metadata_config = metadata_config
+        self.metadata = OpenMetadata(metadata_config)
+        self.service_connection = self.config.serviceConnection.__root__.config
+        self.service = self.metadata.get_service_or_create(
+            entity=DatabaseService, config=config
+        )
+        self.connection = get_connection(self.service_connection)
+        self.client = self.connection.client
+
+    @classmethod
+    def create(cls, config_dict, metadata_config: OpenMetadataConnection):
+        config: WorkflowSource = WorkflowSource.parse_obj(config_dict)
+        connection: DatalakeConnection = config.serviceConnection.__root__.config
+        if not isinstance(connection, DatalakeConnection):
+            raise InvalidSourceException(
+                f"Expected DatalakeConnection, but got {connection}"
+            )
+        return cls(config, metadata_config)
+
+    def prepare(self):
+        pass
+
+    def next_record(self) -> Iterable[Entity]:
+        try:
+
+            bucket_name = self.service_connection.bucketName
+            if hasattr(self.service_connection.configSource, "gcsConfig"):
+                yield from self.get_gcs_files(bucket_name)
+
+            if hasattr(self.service_connection.configSource, "awsConfig"):
+                yield from self.get_s3_files(bucket_name)
+
+        except Exception as err:
+            logger.error(traceback.format_exc())
+            logger.error(err)
+
+    def read_csv_from_gcs(self, key, bucket_name):
```

Comment:
Could we move all these `read_X_from_Y` functions to different modules? such as `s3_utils` or `gcs_utils`